internal/controller/certificates/policies/checks.go

```diff
@@ -355,143 +355,249 @@ func SecretTemplateMismatchesSecret(input Input) (string, string, bool) {
-		// Only attempt to decode the signed certificate, if one is available.
-		var x509cert *x509.Certificate
-		if len(input.Secret.Data[corev1.TLSCertKey]) > 0 {
-			var err error
-			x509cert, err = pki.DecodeX509CertificateBytes(input.Secret.Data[corev1.TLSCertKey])
-			if err != nil {
-				// This case should never happen as it should always be caught by the
-				// secretPublicKeysMatch function beforehand, but handle it just in case.
-				return InvalidCertificate, fmt.Sprintf("Failed to decode stored certificate: %v", err), true
+		managedLabels, managedAnnotations, err := secretLabelsAndAnnotationsManagedFields(input.Secret, fieldManager)
+		if err != nil {
+			return ManagedFieldsParseError, fmt.Sprintf("failed to decode managed fields on Secret: %s", err), true
+		}
+
+		// Remove the non cert-manager annotations from the managed Annotations so we can compare
+		// 1 to 1 all the cert-manager annotations.
+		for k := range managedAnnotations {
+			if strings.HasPrefix(k, "cert-manager.io/") ||
+				strings.HasPrefix(k, "controller.cert-manager.io/") {
+				continue
 			}
+
+			delete(managedAnnotations, k)
 		}
 
-		baseAnnotations, err := internalcertificates.AnnotationsForCertificate(x509cert)
+		// Ignore the CertificateName and IssuerRef annotations as these cannot be set by the postIssuance controller.
+		for _, k := range []string{
+			cmapi.CertificateNameKey,       // SecretCertificateNameAnnotationMismatch checks the value
+			cmapi.IssuerNameAnnotationKey,  // SecretIssuerAnnotationsMismatch checks the value
+			cmapi.IssuerKindAnnotationKey,  // SecretIssuerAnnotationsMismatch checks the value
+			cmapi.IssuerGroupAnnotationKey, // SecretIssuerAnnotationsMismatch checks the value
+		} {
+			delete(managedAnnotations, k)
+		}
+
+		// Remove the non cert-manager labels from the managed Annotations so we can compare
+		// 1 to 1 all the cert-manager labels.
+		for k := range managedLabels {
+			if strings.HasPrefix(k, "cert-manager.io/") ||
+				strings.HasPrefix(k, "controller.cert-manager.io/") {
+				continue
+			}
+
+			delete(managedLabels, k)
+		}
+
+		expCertificateDataAnnotations, err := certificateDataAnnotationsForSecret(input.Secret)
 		if err != nil {
 			return InvalidCertificate, fmt.Sprintf("Failed getting secret annotations: %v", err), true
 		}
 
-		// We don't use the values of these annotations, but we need to make sure
-		// that the keys are present in the map so that we can compare the sets.
-		baseAnnotations[cmapi.CertificateNameKey] = "<certificate-value>"
-		baseAnnotations[cmapi.IssuerNameAnnotationKey] = "<issuer-name>"
-		baseAnnotations[cmapi.IssuerKindAnnotationKey] = "<issuer-kind>"
-		baseAnnotations[cmapi.IssuerGroupAnnotationKey] = "<issuer-group>"
-
-		managedLabels, managedAnnotations := sets.NewString(), sets.NewString()
+		expLabels := sets.New[string](
+			cmapi.PartOfCertManagerControllerLabelKey, // SecretBaseLabelsMismatch checks the value
+		)
```

Comment:
Ah, so that's how you create a generic set! I looked at the code the other day and didn't understand how to use it.